File: apps/live_control_server/models/rules_query.py
"""Read-only, citation-bearing Rules Lawyer query packet."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RulesQueryRequest(StrictModel):
    schema_version: Literal["dmb_rules_query_request_v1"] = "dmb_rules_query_request_v1"
    question: str = Field(min_length=1, max_length=2000)
    ruleset_id: str = Field(min_length=1, max_length=100)
    max_hits: int = Field(default=5, ge=1, le=20)

    @field_validator("question", "ruleset_id")
    @classmethod
    def strip_required(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("must not be blank")
        return stripped


RulesQueryStatus = Literal[
    "success", "no_evidence", "insufficient_evidence",
    "rules_space_unavailable", "downstream_failure",
]


class RulesEvidenceItem(StrictModel):
    rank: int = Field(ge=1)
    entity_id: str
    assertion_id: str
    evidence_ref_id: str
    evidence_unit_id: str
    source_artifact_id: str
    source_revision_id: str | None = None
    source_uri: str | None = None
    source_locator: str | None = None
    locator: str | None = None
    source_anchor_id: str | None = None
    excerpt: str | None = None


class RulesQueryTrace(StrictModel):
    searched_revision_id: str
    search_result_digest: str | None = None
    searched_entities: int = 0
    admitted_assertions: int = 0
    returned_evidence: int = 0
    completeness: Literal["complete", "partial", "unavailable"]
    reason: str | None = None


class RulesQueryPacket(StrictModel):
    schema_version: Literal["dmb_rules_query_packet_v1"] = "dmb_rules_query_packet_v1"
    query_id: str
    ruleset_id: str
    rules_space_id: str | None = None
    rules_revision_id: str | None = None
    status: RulesQueryStatus
    evidence: list[RulesEvidenceItem] = Field(default_factory=list)
    trace: RulesQueryTrace

File: apps/live-control-ui/src/api/rulesQuery.ts
export interface RulesEvidenceItem {
  rank: number;
  entity_id: string;
  assertion_id: string;
  evidence_ref_id: string;
  evidence_unit_id: string;
  source_artifact_id: string;
  source_revision_id: string | null;
  source_uri: string | null;
  source_locator: string | null;
  locator: string | null;
  source_anchor_id: string | null;
  excerpt: string | null;
}

export interface RulesQueryPacket {
  schema_version: "dmb_rules_query_packet_v1";
  query_id: string;
  ruleset_id: string;
  rules_space_id: string | null;
  rules_revision_id: string | null;
  status: "success" | "no_evidence" | "insufficient_evidence" | "rules_space_unavailable" | "downstream_failure";
  evidence: RulesEvidenceItem[];
  trace: {
    completeness: "complete" | "partial" | "unavailable";
    reason: string | null;
  };
}

const baseUrl = (import.meta.env.VITE_LIVE_API_BASE_URL as string | undefined) ?? "";

export async function queryRules(question: string, signal?: AbortSignal): Promise<RulesQueryPacket> {
  const response = await fetch(`${baseUrl}/api/live/rules/query`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      schema_version: "dmb_rules_query_request_v1",
      question,
      ruleset_id: "dnd5e-2024-srd-occupancy-v1",
      max_hits: 5,
    }),
    signal,
  });
  if (!response.ok) throw new Error(`Rules service returned ${response.status}`);
  const packet: unknown = await response.json();
  if (
    typeof packet !== "object" || packet === null ||
    (packet as RulesQueryPacket).schema_version !== "dmb_rules_query_packet_v1" ||
    !Array.isArray((packet as RulesQueryPacket).evidence)
  ) throw new Error("Invalid rules query packet");
  return packet as RulesQueryPacket;
}
